### PR TITLE
Add ability to restore status on selected resources

### PR DESCRIPTION
**Pull request by RafaeLeal**
# Summary

This adds a new field on `Restore`, to allow us to select some resources to restore the status.
The way this works it's that it will have the same `includedResources `/`excludedResources` but under the `restoreStatus` key. The default value will be excluding all resources, maintaining the current behavior

```yaml
spec: 
  restoreStatus:
    includedResources: ["webhooks", "workflows"]
```

# Does your change fix a particular issue?

Fixes https://github.com/vmware-tanzu/velero/issues/1687

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by RafaeLeal** [comment]
Not sure what's going on with DCO, I added the `Signed-off-by` on every commit 😕

**Comment by hamza3202** [comment]
This should be a good fix for [this issue](https://github.com/vmware-tanzu/velero/issues/1687#issuecomment-1070943320)

**Comment by RafaeLeal** [comment]
> Not sure what's going on with DCO, I added the `Signed-off-by` on every commit 😕

It was a typo on my email 🤦

**Comment by reasonerjt** [comment]
Based on a previous comment:
https://github.com/vmware-tanzu/velero/issues/1687#issuecomment-524928609
I'm not sure allowing users to choose to restore the status is a good idea, and whether it may cause unexpected problems.
I hope we can have more discussion before the PR is merged.

**Comment by RafaeLeal** [comment]
> Based on a previous comment: [#1687 (comment)](https://github.com/vmware-tanzu/velero/issues/1687#issuecomment-524928609) I'm not sure allowing users to choose to restore the status is a good idea, and whether it may cause unexpected problems. I hope we can have more discussion before the PR is merged.

Sure, what can I do to move this discussion forward?
I can see that ideally all `status` should be recoverable from the `spec`, but we have noticed multiple scenarios where this is either not true or has other undesirable consequences.

Having CRDs to manage external resources should become more common as people adopt [operator pattern](https://github.com/cncf/tag-app-delivery/blob/main/operator-wg/whitepaper/Operator-WhitePaper_v1-0.md). This reference is often something that you can't recover from the spec.

But even when you are indeed able to restore it, it means that you're going to need to request an external service to do that, and in a backup/restore scenario, you will put a lot of unnecessary load on this external service. In my scenario, the load is not a problem, but since it's a third party we can reach rate limits. 

This is especially bad because this disruption happens in both clusters (the one being backup and the one restoring the backup) since it's outside the cluster.

**Comment by reasonerjt** [comment]
> Having CRDs to manage external resources should become more common as people adopt [operator pattern](https://github.com/cncf/tag-app-delivery/blob/main/operator-wg/whitepaper/Operator-WhitePaper_v1-0.md). This reference is often something that you can't recover from the spec.

I hope we can discuss this in the community meeting with other maintainers.  For example, is `status` the only field you can leverage, can you use `annotation` to store the information you need for restore?

**Comment by blackpiglet** [comment]
Understand the use case. Some CRD may store information of resource in status.
In my opinion, the resources can be specified to restore status should be limited to CRD, because k8s native resources don't need status for replications. 
Maybe more check conditions should be added.

**Comment by RafaeLeal** [comment]
> I hope we can discuss this in the community meeting with other maintainers. For example, is `status` the only field you can leverage, can you use `annotation` to store the information you need for restore?

The problem is not to store the information itself, since Velero already saves the serialized status, but the restore process itself. We could add an annotation before restoring, but that would mean that the controller would have to be changed to read that annotation... and maybe we are running open-source controllers like Argo, Prometheus Operator, or Flagger and we can't just change how those controllers work.

**Comment by RafaeLeal** [comment]
> Understand the use case. Some CRD may store information of resource in status. In my opinion, the resources can be specified to restore status should be limited to CRD, because k8s native resources don't need status for replications. Maybe more check conditions should be added.

Interesting.. we could consider that. Right now I believe the only way to know if a resource is a CRD or not is to reach the CRD Kubernetes API, right?

**Comment by blackpiglet** [comment]
> Right now I believe the only way to know if a resource is a CRD or not is to reach the CRD Kubernetes API, right?
May be not that complicated. k8s resources are located in specific API group, for example, core group, app group, batch group. Others extensions's group ends with "k8s.io".

**Comment by sseago** [comment]
There are other non-core resources other than CRDs. For example, in an OpenShift cluster, imagestreams are openshift-specific resources, but if you look at the list of CRDs in an openshift cluster, imagestreams won't be in the list. Point being that if you're looking for "resources that are not core k8s resources", looking at the CRDs API is probably the wrong way to go about it. If the idea is to restore status for non-core resources only, maybe a whitelist of specific core API groups would work, though.

**Comment by RafaeLeal** [comment]
Two things I'd like to add up:
- This implementation, by default, doesn't restore status of *any* resource. You need to explicitly say which resources you want to restore. I think this prevents a lot of issues we could have.
- In k8s, you can have resources in multiple api groups. So looking at those to see if it's "native resource" could lead to some inconsistent behaviour like restoring status for some instances and not all of them.

**Comment by reasonerjt** [comment]
In today's community discussion we concluded that velero should deal with this problem despite storing the reference in `status` field may be an anti-pattern.

We'll check this PR in more details and try to merge it in v1.9

File: pkg/restore/restore.go
Comment (shawn-hurley) [inline pkg/restore/restore.go:1329]: I think that you will need to look up if the resource supports the status subresource otherwise for some resources this will return a 404 not found.

Comment (RafaeLeal) [reply pkg/restore/restore.go:1329]: I was thinking about this problem...
Ideally, we would call the discovery API to know if the resource supports status subresource without coupling with CRDs. 
We could, however, just ignore 404 errors, and maybe log a warning. This way we don't add a new call to the API, and also don't change the RBAC permissions needed to run velero's server. WDYT?

File: pkg/restore/restore.go
Comment (Lyndon-Li) [inline pkg/restore/restore.go:1130]: It's better put this reset further below. As you can see, when processing RestoreItemAction, the obj may be replaced with the return value of the item action, so the reset to obj here may be overwritten. Maybe we can put it right above resourceClient.Create(obj)?

File: pkg/restore/restore.go
Comment (Lyndon-Li) [inline pkg/restore/restore.go:1266]: As far as my understanding: since the status has been removed from "fromCluster" here, but the status in "obj" is not removed when shouldRestoreStatus = true, this will make equality.Semantic.DeepEqual(fromCluster, obj)@1279 returns false all the time, which changes the original logic. Am I wrong?

File: pkg/restore/restore.go
Comment (RafaeLeal) [reply pkg/restore/restore.go:1130]: I was trying to change as little as possible, but sure, we can do that :)

File: pkg/restore/restore.go
Comment (RafaeLeal) [reply pkg/restore/restore.go:1266]: This part of the code is doing the same thing, just split the two tasks...
But I believe you're right. In the scenario that you're restoring status (shouldRestoreStatus = true), and the resource already exists, as is, it's always triggering the update.
Let me try to fix it

Comment (RafaeLeal) [reply pkg/restore/restore.go:1282]: I just pushed a fix for this, let me know what you think

**Comment by blackpiglet** [comment]
@RafaeLeal 
There is an error in CI action.
`CRD verification - failed! Generated CRDs are out-of-date, please run 'make update' and 'git add' the generated file(s).`

Please run `make update` to generate CRD's deep-copy functions, then commit again.

File: pkg/restore/restore.go
Comment (Lyndon-Li) [inline pkg/restore/restore.go:1386]: Comment 5: set the status value of obj from the objWithStatus or the kept the status value

File: pkg/restore/restore.go
Comment (RafaeLeal) [reply pkg/restore/restore.go:1130]: I submitted a PR with these suggestions on the forked repository
https://github.com/RafaeLeal/velero/pull/1/files

I feel that's still complex, but it's hard to simplify without major refactorings.
In this suggested approach, you need to add and remove fields into an unstructured object, which I was trying to avoid... But I guess it's less intrusive. Let me know if that's what you intended

Comment (Lyndon-Li) [reply pkg/restore/restore.go:1130]: "I guess it's less intrusive".  --- yes, agree. This is important I think, because we can avoid regressions.
"need to add and remove fields into an unstructured object" --- if you don't want to do this, I guess you can still copy the entire object but use the status field only. Anyway, let me review this with the team and get you back.
"I feel that's still complex". --- somehow better IMO, let me review with the team for any further improvement

Thanks a lot for your efforts.

Comment (RafaeLeal) [reply pkg/restore/restore.go:1130]: I have little context about this.. so I'm comfortable with both solutions, so I'd go with whatever the team feels it's better to maintain.
> This is important I think, because we can avoid regressions.
I think that if we are afraid of doing more intrusive changes, we should refactor the code so we can have better test coverage. I know that's an "easier said than done" kind of issue... but maybe we can come up with something

Comment (Lyndon-Li) [reply pkg/restore/restore.go:1130]: @RafaeLeal 
Please help to merge https://github.com/RafaeLeal/velero/pull/1/files, since this make less change and less intrusive.
For the overall refactor to the existing logics, we may visit it during the data mover design & change, let's keep the existing logic as is for this PR.

@reasonerjt 
Please also help to have a look at the change where unstructured.NestedFieldCopy，unstructured.SetNestedField are used, see if you have any other proposal


Thanks.

Comment (RafaeLeal) [reply pkg/restore/restore.go:1130]: Merged 👍

File: pkg/restore/restore.go
Comment (Lyndon-Li) [inline pkg/restore/restore.go:1012]: This line seems not necessary as this code block doesn't really create the object. And there is one call resetMetadataAndStatus at line 1133 later before creating the object

File: pkg/restore/restore.go
Comment (reasonerjt) [inline pkg/restore/restore.go:1697]: It's always nil, why return this error?

File: pkg/restore/restore.go
Comment (reasonerjt) [reply pkg/restore/restore.go:1381]: @RafaeLeal if that's the case, does it mean this feature only works for resources whose status is subresource?  Is that an acceptable limitation?

In addition I do we have enough information in the `err` ?  LIke the name of the resources?

File: pkg/restore/restore.go
Comment (Lyndon-Li) [reply pkg/restore/restore.go:1012]: Sorry for any misleading, but by the last comment I meant the line at 1012, resetStatus(obj), is not necessary, so please consider and remove it. Thanks.

File: pkg/restore/restore.go
Comment (reasonerjt) [reply pkg/restore/restore.go:1130]: I think it look good.

File: pkg/restore/restore.go
Comment (Lyndon-Li) [inline pkg/restore/restore.go:1012]: Sorry for any misleading, but by the last comment I meant the line at 1012, resetStatus(obj), is not necessary, so please consider and remove it. Thanks.

File: pkg/cmd/cli/restore/create.go
Comment (blackpiglet) [inline pkg/cmd/cli/restore/create.go:286]: Should o.StatusExcludeResources length be checked too?

Comment (blackpiglet) [inline pkg/cmd/cli/restore/create.go:286]: Should o.StatusExcludeResources length be checked, too?

File: pkg/restore/restore.go
Comment (blackpiglet) [reply pkg/restore/restore.go:1381]: Looks like this is not related to whether the status of resource is defined as sub-resource. 
To me, as long as the resource's client implements the UpdateStatus method, then it will work.

**Comment by reasonerjt** [comment]
@RafaeLeal please also update the doc to include the change you added to the CRD:
https://github.com/vmware-tanzu/velero/blob/bfdb68a35a51e86ca8839567d15a63de8a673201/site/content/docs/main/api-types/restore.md

File: pkg/restore/restore.go
Comment (RafaeLeal) [reply pkg/restore/restore.go:1697]: This was part of the old code that I just splitted the `resetMetadataAndStatus` function.. 
But sure, I believe it's better to use the unstructured API :)

File: pkg/cmd/cli/restore/create.go
Comment (RafaeLeal) [reply pkg/cmd/cli/restore/create.go:286]: Well, I believe if you don't include any resources, it won't trigger anything anyway

File: pkg/restore/restore.go
Comment (RafaeLeal) [reply pkg/restore/restore.go:1380]: @reasonerjt I think this is acceptable.. For me at least it seems that would be super weird to have a CRD with a "status" field that is not a subresource "status".

I believe that we have very raw info on `err`, basically the HTTP response from a 404 on Kubernetes. Which is something like
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "deployments.apps \"dont-exist\" not found",
  "reason": "NotFound",
  "details": {
    "name": "dont-exist",
    "group": "apps",
    "kind": "deployments"
  },
  "code": 404
  ```
  
  @blackpiglet since we are using the dynamic client, it will always implement the `UpdateStatus` method. The problem is the server answer itself

**Review by reasonerjt — APPROVED** [review]

**Review by blackpiglet — APPROVED** [review]

**Pull request merged**
